src/core/meshcop/tcat_agent.cpp

```diff
@@ -629,13 +656,134 @@ Error TcatAgent::HandleGetProvisioningUrl(Message &aOutgoingMessage, bool &respo
     length = StringLength(mVendorInfo->mProvisioningUrl, kProvisioningUrlMaxLength);
     VerifyOrExit(length > 0 && length <= Tlv::kBaseTlvMaxLength, error = kErrorInvalidState);
 
-    error    = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, mVendorInfo->mProvisioningUrl, length);
+    SuccessOrExit(error =
+                      Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, mVendorInfo->mProvisioningUrl, length));
+    response = true;
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentPskdHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error                    error = kErrorNone;
+    Crypto::HmacSha256::Hash hash;
+    Crypto::HmacSha256::Hash recivedHash;
+
+    VerifyOrExit(mRandomChallenge != 0, error = kErrorFailed);
+    VerifyOrExit(aLength == Crypto::HmacSha256::Hash::kSize, error = kErrorSecurity);
+
+    SuccessOrExit(error = aIncomingMessage.Read(aOffset, recivedHash.m8, aLength));
+
+    hash = CalculateHash(mRandomChallenge, mVendorInfo->mPskdString,
+                         StringLength(mVendorInfo->mPskdString, kMaxPSKdLength));
+
+    VerifyOrExit(hash == recivedHash, error = kErrorSecurity);
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentPskcHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error                    error = kErrorNone;
+    Crypto::HmacSha256       hmac;
+    Crypto::HmacSha256::Hash hash;
+    Crypto::HmacSha256::Hash recivedHash;
+    Dataset::Info            datasetInfo;
+    Pskc                     pskc;
+
+    VerifyOrExit(Get<ActiveDatasetManager>().Read(datasetInfo) == kErrorNone, error = kErrorSecurity);
+    VerifyOrExit(datasetInfo.IsPresent<Dataset::kPskc>(), error = kErrorSecurity);
```

Comment:
I'm wondering if in case of failure these two `VerifyOrExit` should return `kErrorSecurity` ? It seems misaligned with later `VerifyOrExit(mRandomChallenge != 0, error = kErrorFailed);`

Reply:
I have aligned all errors to `kErrorSecurity`.